#513 multiple diverse drafts

benjaminking opened the pull request:
This adds the ability for SILNLP's translate functionality to produce multiple diverse drafts.  This ability is controlled by the --multiple-translations parameter on the command line (for both translate.py and experiment.py).  The other parameters, such as number of drafts and the method of creating multiple translations, are specified in the config file.

When multiple translations are requested, the output file name is given a different extension for each draft.  For example, if you specify "output.txt" as the output file and request 3 drafts, then three files will be created called "output.1.txt", "output.2.txt", and "output.3.txt".

The default method for producing multiple translations is called "hybrid" and uses a combination of beam search and sampling.  Other supported values are "sampling" which creates each draft with random sampling, "beam_search" which uses vanilla beam search (using the top-n hypotheses to populate the n drafts), and "diverse_beam_search" which uses the method in Vijayakumar, et al., 2016.  Other new parameters controlling the creation of multiple drafts are "temperature" and "diversity_penalty".  I will plan to document all of this in the wiki.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/sillsdev/silnlp/533)
<!-- Reviewable:end -->

ddaspit reviewed (changes requested):
> Reviewed 8 of 8 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/silnlp/533)* status: all files reviewed, 6 unresolved discussions (waiting on @benjaminking)

___
*[`silnlp/common/translator.py` line 50 at r1](https://reviewable.io/reviews/sillsdev/silnlp/533#-O7p5pgE4Nwr1-bxOtl7:-O7p5pgE4Nwr1-bxOtl8:b-gm6re8) ([raw file](https://github.com/sillsdev/silnlp/blob/b7b5419d79e0e714a331b6d5b106b1626ef63e9e/silnlp/common/translator.py#L50)):*
> ```Python
> 
> # A group of multiple translations of a single sentence
> class TranslationGroup:
> ```

`TranslationGroup` and `TranslatedDraft` are both simple wrappers of collection classes without any added functionality. I am unsure about the purpose of these classes. Is this just a form of type aliasing to give the types more descriptive names? If so, you can do something like this:
``` py
TranslationGroup = List[str]
```
___
*[`silnlp/common/translator.py` line 60 at r1](https://reviewable.io/reviews/sillsdev/silnlp/533#-O7p7QTP3p_TcZyN77Rv:-O7p7QTP3p_TcZyN77Rw:bgm0mko) ([raw file](https://github.com/sillsdev/silnlp/blob/b7b5419d79e0e714a331b6d5b106b1626ef63e9e/silnlp/common/translator.py#L60)):*
> ```Python
> # An iterable representing a single draft (one translation of each input sentence)
> class TranslatedDraft:
>     def __init__(self, sentences: Iterable[str]):
> ```

`sentences` could be iterated multiple times. This can result in unexpected behavior in some cases. It would be safer to pass in a list.
___
*[`silnlp/common/translator.py` line 67 at r1](https://reviewable.io/reviews/sillsdev/silnlp/533#-O7p5ZV5CmW8cD-hHIup:-O7p5ZV5CmW8cD-hHIuq:b-ovl7pk) ([raw file](https://github.com/sillsdev/silnlp/blob/b7b5419d79e0e714a331b6d5b106b1626ef63e9e/silnlp/common/translator.py#L67)):*
> ```Python
> # list of translation groups as a collection of discrete drafts
> class DraftGroup:
>     def __init__(self, translation_groups: Iterable[TranslationGroup]):
> ```

`translation_groups` is iterated multiple times. This can result in unexpected behavior in some cases. It would be safer to pass in a list.
___
*[`silnlp/common/translator.py` line 71 at r1](https://reviewable.io/reviews/sillsdev/silnlp/533#-O7p6OFZAqS8cjJGgJB1:-O7p6OFZAqS8cjJGgJB2:b-vf8x12) ([raw file](https://github.com/sillsdev/silnlp/blob/b7b5419d79e0e714a331b6d5b106b1626ef63e9e/silnlp/common/translator.py#L71)):*
> ```Python
>         self._calculate_num_drafts()
> 
>     def _calculate_num_drafts(self):
> ```

You missed the return type here. Mypy won't perform type checking without it.
___
*[`silnlp/nmt/hugging_face_config.py` line 309 at r1](https://reviewable.io/reviews/sillsdev/silnlp/533#-O7own2C9z2KCYg82nxz:-O7own2C9z2KCYg82ny-:b82uubv) ([raw file](https://github.com/sillsdev/silnlp/blob/b7b5419d79e0e714a331b6d5b106b1626ef63e9e/silnlp/nmt/hugging_face_config.py#L309)):*
> ```Python
>                 "infer": {
>                     "infer_batch_size": 16,
>                     "num_beams": 5,
> ```

This will change the default behavior when performing standard inferencing. Could we have a separate `num_beams` setting that is used when generating multiple translations. 
___
*[`silnlp/nmt/test.py` line 416 at r1](https://reviewable.io/reviews/sillsdev/silnlp/533#-O7p-AQz4XKO-Kkxt-Xl:-O7p-AQz4XKO-Kkxt-Xm:b6dh2b5) ([raw file](https://github.com/sillsdev/silnlp/blob/b7b5419d79e0e714a331b6d5b106b1626ef63e9e/silnlp/nmt/test.py#L416)):*
> ```Python
>             source_paths,
>             translation_paths,
>             False,
> ```

It makes the code more readable to explicitly specify the parameter name when passing a literal, i.e. `produce_multiple_translations=False`.


<!-- Sent from Reviewable.io -->

benjaminking commented:
I have addressed these issues with the latest commit.  I reverted "num_beams" back to 2.  My evaluations had found that a value of 5 led to higher BLEU scores, but I will leave it up to the user to make that change.

ddaspit reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/silnlp/533#-:-O83AQ1t14t5D1dTLjsi:bnfp4nl)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @benjaminking)



<!-- Sent from Reviewable.io -->
